internal/leadership/elector.go

```diff
@@ -114,8 +114,18 @@ func (e *Elector) Run(ctx context.Context) {
 		}
 	}
 
-	subscription := e.notifier.Listen(notifier.NotificationTopicLeadership, handleNotification)
-	defer subscription.Unlisten()
+	sub, err := e.notifier.Listen(ctx, notifier.NotificationTopicLeadership, handleNotification)
+	if err != nil {
+		if !errors.Is(err, context.Canceled) {
+			// TODO(brandur): We shouldn't be silently swallowing major start up
+			// problems like this. Unfortunately, Run doesn't return an error, so
+			// fixing this will be a big refactor, so don't do it right now.
+			e.logger.Error(e.name+": Error listening for leadership changes", "err", err, "topic", notifier.NotificationTopicJobControl)
```

Comment:
See TODO here, but I don't like this and will fix it, but can't right now. I'd really like to see both the elector and producer return errors when problems happen on start up instead of silently swallowing errors.

It'll need some refactoring to get there, but once I do that this will return an error instead of producing a log statement with a TODO.

Reply:
> I think the challenge is that it's probably desirable to have the client fail to start if any of its critical subprocesses fail to start correctly at initial boot, however if those subprocesses fail at any other point they should enter into a graceful exponential backoff loop to try and get healthy again.

Ah yeah, as mentioned on Slack, this is exactly how the new notifier is written to work: we fail fast in case of a major problem so the user can find out about it, but subsequent failures after start are in internalized, and services will do their best to make sure they get healthy again. We also now have tests in the notifier to prove that this works, helping to take the guesswork out of it. (See `WaitErrorAndBackoff`, `BackoffSleepCancelledOnStop`, and `StillFunctionalAfterMainLoopFailure`.)